pkgs/servers/nosql/janusgraph/wrapper.nix

```
@@ -0,0 +1,53 @@
+{ janusgraph-unwrapped
+, stdenv
+, lib
+, fetchzip
+, callPackage
+, jdk11
+, makeWrapper
+}:
+
+stdenv.mkDerivation {
+  inherit (janusgraph-unwrapped) pname version;
+
+  src = janusgraph-unwrapped;
```

```suggestion
  inherit (janusgraph-unwrapped) src;
```
You likely meant this, right?

---

I did not. The source of the unwrapped derivation may be either Java source code, or binary  jars from the release tarball. The derivation builds source code to jars, and copies the expected outputs to $out, which is then used as $src in the wrapper.

---

Why do we need both variants? If we can build from source then we don't need the binary release.

---

You're right, it is now possible to replace the binary zip download with a GitHub source fetch at the version tag. But I figured, since an official binary release exists, we might as well use it; if the source build ever breaks, the binary release will keep working.

The rationale for exposing both unwrapped and wrapped versions is for debugging purposes. Building the unwrapped package should be the equivalent of downloading a binary release. The unwrapped package is the `default.nix` that I wish was present upstream. It is also a fallback in case the wrapper fails, or is not desired, e.g. on platforms other than NixOS.

---

> But I figured, since an official binary release exists, we might as well use it; if the source build ever breaks, the binary release will keep working.

This is not the distro spirit we want to follow. If we can compile something feasible from  from source then we should do it and don't need to offer a binary variant. We always have the cache backing us up and people can always download the binary from upstream.

We also do not normally differentiate packages between NixOS and none NixOS, packages should behave and be build the same way to keep things unified.

---

I see. Thanks for the explanation.

Unfortunately, JanusGraph fails to build on the v0.6.2 tag, with some Java errors. I'm not sure if spending effort patching it is a good use of our effort.

Would it be better if I called the from-binary package `janusgraph-bin`?

---

> I'm not sure if patching an already outdated build is a good use of our time and effort at this point.

I don't know java but one patch release should be fine.



> But I still think it is valuable to offer the latest official release. Would it be better if I called the from-binary package `janusgraph-bin`?

yes, sure but we do not need both. Please choose one.

---

I think it can make sense to provide both a stable and an unstable version if there is a demand for both. In this case the latest stable version is already ~7 months old, but it is a stable release. Master is of course recent with a presumably a lot of improvements but not yet stable. I don't know janusgraph/hugegraph enough to know how many users there are and whether there's users for both versions though.

Alternatively though, what if we just package the new 1.0.0-rc1 release candidate released [just last week](https://github.com/JanusGraph/janusgraph/releases/tag/v1.0.0-rc1)? I also expect the full release to be soon

---

I've removed the `-unstable` versions, as it is now easy enough to override the source for the stable packages. Both packages build from source, but it is possible to override the `-unwrapped` package to point it to a binary release.